File: tests/tile.py
# Copyright 2021 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import cunumeric as lg


def test():
    a = lg.array([0, 1, 2])

    b = lg.tile(a, 4)
    assert lg.array_equal(b, [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2])

    c = lg.tile(a, (3, 4))
    assert lg.array_equal(
        c,
        [
            [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2],
            [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2],
            [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2],
        ],
    )

    d = lg.tile(a, (3, 1, 4))
    assert lg.array_equal(
        d,
        [
            [[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]],
            [[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]],
            [[0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]],
        ],
    )

    e = lg.array([[1, 2], [3, 4]])

    f = lg.tile(e, 2)
    assert lg.array_equal(f, [[1, 2, 1, 2], [3, 4, 3, 4]])

    g = lg.tile(e, (2, 1))
    assert lg.array_equal(g, [[1, 2], [3, 4], [1, 2], [3, 4]])

    return


if __name__ == "__main__":
    test()

File: src/cunumeric/matrix/tile.cc
/* Copyright 2021-2022 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "cunumeric/matrix/tile.h"
#include "cunumeric/matrix/tile_template.inl"

namespace cunumeric {

using namespace legate;

template <typename VAL, int32_t OUT_DIM, int32_t IN_DIM>
struct TileImplBody<VariantKind::CPU, VAL, OUT_DIM, IN_DIM> {
  void operator()(const Rect<OUT_DIM>& out_rect,
                  const Pitches<OUT_DIM - 1>& out_pitches,
                  size_t out_volume,
                  const Point<IN_DIM>& in_strides,
                  const AccessorWO<VAL, OUT_DIM>& out,
                  const AccessorRO<VAL, IN_DIM>& in) const
  {
    for (size_t out_idx = 0; out_idx < out_volume; ++out_idx) {
      const auto out_point = out_pitches.unflatten(out_idx, out_rect.lo);
      const auto in_point  = get_tile_point(out_point, in_strides);
      out[out_point]       = in[in_point];
    }
  }
};

/*static*/ void TileTask::cpu_variant(TaskContext& context)
{
  tile_template<VariantKind::CPU>(context);
}

namespace  // unnamed
{
static void __attribute__((constructor)) register_tasks(void) { TileTask::register_variants(); }
}  // namespace

}  // namespace cunumeric
